### PR TITLE
[UXIT-1523] Ecosystem Explorer Project Update Form [skip percy]

### PR DESCRIPTION
**Pull request by CharlyMartin**
## 📝 Description

This PR adds an update step to the Ecosystem Explorer Update Form

## 🛠️ Key Changes

- Creates an `actions` folder for all [server actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) used in the `useCreateEcosystemProjectForm` and `useUpdateEcosystemProjectForm` hooks.
- Implements `useSWR` to manage the state of async action functions in client components. It also introduces `SWR_KEYS` to manage the [query keys](https://swr.vercel.app/docs/api#parameters).
- Adds `submitProjectWithLogo`, which is used when creating a project or updating a project with a new logo.
- Adds `submitProjectWithoutLogo,` which is only used when project information is updated, but the logo remains unchanged.
- Updates `getInitialFormData` to accept existing project data and format it according to the form data structure.
- Simplifies `FormFileInput` to only accept a `File`, instead of `Array<File>`

## 🧪 How to Test

1. Go to `/ecosystem-explorer/project-form`
2. Select a project to update
3. Submit and inspect the open PR.

## 📸 Screenshots

CleanShot 2024-10-16 at 09 36 29

## 🔖 Resources

- [NUQS](https://nuqs.47ng.com/) - to sync `useState` with the URL query.
- [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) - to call server-side functions in the client without creating API endpoints
- [useSWR](https://swr.vercel.app/docs/api)

### CODE REVIEW

**Comment by barbaraperic** [comment]
Maybe we can change the title to `Ecosystem Project Form Update`?
Screenshot 2024-10-14 at 12 10 02

**Comment by barbaraperic** [comment]
Sometimes the year resets without toggling the select Filecoin or IPPF above

Screenshot 2024-10-14 at 12 14 52

**Comment by barbaraperic** [comment]
The SEO fields change? I guess it has to do with our recent refactor with Open Graph?

Screenshot 2024-10-14 at 11 46 09

**Comment by barbaraperic** [comment]
I think that sometimes the images are not retrieved (also some project don't have a logo :( ) so there's an error when submitting the project, and when you close the modal you're redirect to `ecosystem-explorer/project-form`. I can't tell which field is giving the error because of the redirection, but I assume it's the image?

Screenshot 2024-10-14 at 12 18 08

Screenshot 2024-10-14 at 12 19 56

File: src/app/ecosystem-explorer/project-form/actions/getProjectsData.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+'use server'
+
+import { getEcosystemProjectsData } from '@/utils/getEcosystemProjectData'
+
+export async function getProjectsData() {
```
Comment (CharlyMartin) [inline src/app/ecosystem-explorer/project-form/actions/getProjectsData.ts:5]: rename to `getAllProjectsData`

File: src/app/_components/Form/FormFileInput/FormFileInput.tsx [hunk 2]
```diff
@@ -37,26 +37,21 @@ export function FormFileInput({
   accept,
   ...rest
 }: FormFileInputProps) {
-  const loadedFile = files?.[0]
-
   function resetFiles() {
```
Comment (mirhamasala) [inline src/app/_components/Form/FormFileInput/FormFileInput.tsx:40]: We are working with a single file now, and yet, we have the functions `resetFiles` and `loadFiles`? Seems confusing?

File: src/app/_components/Form/FormTextarea/FormTextarea.tsx [hunk 3]
```diff
@@ -39,6 +39,9 @@ export function FormTextarea({
         className={clsx(
           'block min-h-44 w-full rounded-lg border border-brand-300 bg-brand-800 px-3.5 py-3 focus:brand-outline placeholder:text-brand-300 hover:border-brand-400 placeholder:hover:text-brand-400 focus:text-brand-100 placeholder:focus:text-brand-100 data-[disabled]:cursor-not-allowed',
           error && 'border-red-400',
+          characterCount > 400 && 'min-h-60',
```
Comment (mirhamasala) [inline src/app/_components/Form/FormTextarea/FormTextarea.tsx:42]: ❤️

File: src/app/ecosystem-explorer/page.tsx [hunk 4]
```diff
@@ -104,10 +109,7 @@ export default function EcosystemExplorer({ searchParams }: Props) {
         title={header.title}
         description={header.description}
         image={graphicsData.ecosystem}
-        cta={{
-          text: 'Submit Your Project',
-          href: PATHS.ECOSYSTEM_EXPLORER_PROJECT_FORM.path,
-        }}
+        cta={cta}
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/page.tsx:112]: Can we revert this? We are using it everywhere. 🙏🏼

File: src/app/ecosystem-explorer/page.tsx [hunk 5]
```diff
@@ -204,10 +206,7 @@ export default function EcosystemExplorer({ searchParams }: Props) {
       <CTASection
         title="Become Part of the Expanding Ecosystem"
         description="If you’re building on Filecoin and don’t see your project or want to edit your listing, share your details."
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/page.tsx:208]: We should ask @jordanlynnf to look at this copy or adapt it ourselves. It reads weird.

File: src/app/ecosystem-explorer/project-form/components/FormSection.tsx [hunk 6]
```diff
@@ -15,7 +15,12 @@ export function FormSection({
       <Heading tag="h3" variant="xl">
         {title}
       </Heading>
-      {description && <p className="mt-2">{description}</p>}
+
+      {description && (
+        <div className="mt-2">
+          {typeof description === 'string' ? <p>{description}</p> : description}
+        </div>
+      )}
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/components/FormSection.tsx:23]: What use cases does this handle?

File: src/app/ecosystem-explorer/project-form/components/Loader.tsx [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export function Loader() {
+  return <p>Loading...</p>
+}
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/components/Loader.tsx:3]: Not a blocker - Can we add ticket for Filipa?

File: src/app/events/[slug]/components/ScheduleTabs.tsx [hunk 8]
```diff
@@ -94,8 +94,3 @@ export function ScheduleTabs({ schedule }: ScheduleTabsProps) {
     </TabGroup>
   )
 }
-
-function isScreenBelowLg() {
-  return window.matchMedia(`(max-width: ${parseInt(screens.md, 10) - 1}px)`)
-    .matches
-}
```
Comment (mirhamasala) [inline src/app/events/[slug]/components/ScheduleTabs.tsx:101]: Mistake?

File: src/content/pages/ecosystem-explorer/project-form.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
 ---
 header:
-  title: "Add a Project"
+  title: "Add or Update a Project"
```
[the remark this answers is not in the document]
Comment (mirhamasala) [reply src/content/pages/ecosystem-explorer/project-form.md:3]: Yes - We can change it to "Submit or Update a Project" - The GH suggestion is incomplete, by the way.

File: src/app/ecosystem-explorer/project-form/hooks/useCreateEcosystemProjectForm.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+      })
+
+      updateSearchParams({ status: 'success', prNumber: pullRequest.number })
+    } catch (error) {
+      console.error('useCreateEcosystemProjectForm', error)
+      Sentry.captureException(error)
+      updateSearchParams({
+        status: 'error',
+        message:
+          "We couldn't submit your information. Please try again or email us at info@fil.org",
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/hooks/useCreateEcosystemProjectForm.ts:35]: siteMetadata?

File: src/app/ecosystem-explorer/project-form/components/ErrorMessage.tsx [hunk 11]
```diff
@@ -0,0 +1,9 @@
+type ErrorMessageProps = {
+  message: string
+}
+
+export function ErrorMessage({ message }: ErrorMessageProps) {
+  console.error(message)
+
+  return <p>Error: {message}</p>
+}
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/components/ErrorMessage.tsx:9]: Not a blocker but let's create a ticket for Filipa.

File: src/app/ecosystem-explorer/project-form/components/CreateOrUpdateRouter.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+const createOption = actionOptions[0]
+
+export function CreateOrUpdateRouter() {
+  const [action, setAction] = useQueryState(
+    URL_QUERY_KEY,
+    parseAsString.withDefault(createOption.id),
+  )
+
+  return (
+    <div className="md:max-w-readable">
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/components/CreateOrUpdateRouter.tsx:27]: To be moved to the parent.

File: src/app/ecosystem-explorer/project-form/components/CreateOrUpdateRouter.tsx [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+      >
+        <FormRadioGroup
+          options={actionOptions}
+          value={actionOptions.find((option) => option.id === action)}
+          onChange={(option) => setAction(option.id)}
+        />
+      </FormSection>
+
+      {action === 'create' && <EcosystemProjectCreateForm />}
+      {action === 'update' && <SelectProjectToUpdate />}
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/components/CreateOrUpdateRouter.tsx:40]: `ProjectSelector`?

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx [hunk 14]
```diff
@@ -142,25 +170,25 @@ export function EcosystemProjectForm({
...
               label="Category"
-              placeholder="Select category"
+              placeholder="Select Category"
               options={categoryOptions}
               disabled={isSubmitting}
             />
           </div>
           <div className="min-w-0 sm:w-1/2">
             <ControlledFormListbox<EcosystemProjectFormData>
               name="topic"
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx:180]: TBD: Can we change this to subcategory?

File: src/app/ecosystem-explorer/project-form/hooks/useUpdateEcosystemProjectForm.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+      })
+
+      updateSearchParams({ status: 'success', prNumber: pullRequest.number })
+    } catch (error) {
+      console.error('useUpdateEcosystemProjectForm', error)
+      Sentry.captureException(error)
+      updateSearchParams({
+        status: 'error',
+        message:
+          "We couldn't submit your information. Please try again or email us at info@fil.org",
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/hooks/useUpdateEcosystemProjectForm.ts:66]: Same https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/693/files#r1803988809

File: src/app/ecosystem-explorer/project-form/utils/fileUtils.ts [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    reader.onload = () => {
+      const rawResult = String(reader.result)
+      const [, cleanBase64String] = rawResult.split('base64,')
+      resolve(cleanBase64String)
+    }
+    reader.onerror = (error) => reject(error)
+  })
+}
+
+const ALLOWED_IMAGE_FORMATS = ['png', 'jpg', 'jpeg', 'svg', 'webp'] as const
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/project-form/utils/fileUtils.ts:41]: Should we use this in `ControlledFormFileInput`

**Review by mirhamasala — CHANGES_REQUESTED** [review]
@CharlyMartin - I did a first pass at this and also pushed a [CR](hhttps://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/693#pullrequestreview-2374686219) (pretty much all nits but it was quicker this way) - Maybe you can have a look at some of the comments pending (I addressed a few myself - You can probably still find them in resovled), and then we should schedule a pairing. I will need a walkthrough! Thank you.

- [ ] Look into syncing types between form data and EcosystemProject type
- [ ] Under “How do you use Filecoin Network” - Tell the user that they can use Markdown
- [ ] Check if we could merge `submitProjectWithoutLogo` and `SubmitProjectWithLogo`
- [ ] Get @Filipa Ribeiro to look into Loader and ErrorMessage (not blocker)
- [ ] Re-use new `Listbox` components

**Comment by CharlyMartin** [comment]
with [useSWR](https://swr.vercel.app/docs/prefetching) - prefetch data that doesn't have any dependency so that the loading state doesn't create a visual glitch

File: src/app/ecosystem-explorer/project-form/components/CreateOrUpdateRouter.tsx (hunk 13 above)
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/project-form/components/CreateOrUpdateRouter.tsx:40]: ✅

File: src/app/ecosystem-explorer/project-form/utils/fileUtils.ts (hunk 16 above)
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/project-form/utils/fileUtils.ts]: ✅

File: src/app/ecosystem-explorer/project-form/hooks/useCreateEcosystemProjectForm.ts (hunk 10 above)
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/project-form/hooks/useCreateEcosystemProjectForm.ts:35]: https://filecoinfoundation.slack.com/archives/C06MTEQ3SP6/p1726233603097969?thread_ts=1726069025.287869&cid=C06MTEQ3SP6

**Comment by CharlyMartin** [comment]
Closed as we're parking this feature for now. May re-open later.

**Pull request closed without merging**